### ISSUE TITLE
[Python] Restructure Literal and Expression

### ISSUE THREAD

**Issue by dramaticlly** [apache/iceberg#4435]
for context :https://github.com/apache/iceberg/pull/4262#pullrequestreview-899679092

Current python src structure

```
src
├── iceberg
│ ├── __init__.py
│ ├── expression
│ │ ├── __init__.py
│ │ └── literals.py
│ ├── expressions.py
│ ├── io
│ │ ├── __init__.py
│ │ ├── base.py
│ │ └── pyarrow.py
│ ├── types.py
│ └── utils
│     ├── __init__.py
│     └── bin_packing.py
```

@samredai proposed following new structure 

```
src
├── iceberg
│ ├── __init__.py
│ ├── expressions
│ │ ├── __init__.py
│ │ ├── operations.py (renamed from expressions.py)
│ │ └── literals.py
│ ├── io
│ │ ├── __init__.py
│ │ ├── base.py
│ │ └── pyarrow.py
│ ├── types.py
│ └── utils
│     ├── __init__.py
│     └── bin_packing.py
```

The PR change shall be mechanical refactoring and introduce no logical code change

**Comment by samredai** [comment]
Thanks for opening this @dramaticlly, to add more, I think we should section off the ABC classes to a `base.py` file (doesn't have to be the same PR as the general restructuring). Mainly because the expressions module has a pretty gnarly sequence of class inheritance that's hard to follow when the interfaces are scattered among the concrete classes. This also keeps the files with the actual logic easy to follow since the top of those files are not full of super classes. In turn that also makes it easier to find where pdb breakpoints need to be set while debugging.

Here's what I imagine the file contents would look like:
- **base.py** (ABC classes that are never instantiated directly)
  - Literal
  - Expression
  - ExpressionVisitor
  - BoundExpressionVisitor
  - Evaluator
  - Bound
  - Unbound
  - Term
  - Predicate
- **literals.py**
  - IntegerLiteral
  - BooleanLiteral
  - DecimalLiteral
  - ...
- **operation.py**
  - Operation
  - And
  - Or
  - Not
  - TrueExp
  - FalseExp
- **predicate.py**
  - BoundPredicate
  - UnboundPredicate
- **references.py**
  - Boundreference
  - UnboundReference
- **visitors.py**
  - EvalVisitor
  - BindVisitor
  - ReferenceVisitor
  - ManifestEvalVisitor
  - MetricsEvalVisitor
- **evaluators.py**
  - InclusiveManifestEvaluator
  - StrictMetricsEvaluator

I may have left out a couple classes/functions but I think it captures the gist. There's of course a good amount of opinion in this but personally it helps navigate a pretty hefty expressions module. I think of it as "I need to look into an evaluator, must be in evaluator.py" or "I need to debug a visitor, it must be in visitors.py". If I want to get a better understanding of the interfaces and how they all come together, I go to `base.py` and find a collection of well-documented ABC classes.

**Comment by rdblue** [comment]
That seems like a lot of files to me, and some of them only have a few classes. I'd probably move the classes that are expressions into the base or have one file instead of operation.py, predicate.py, and references.py. Visitors I think should probably be their own thing, but I'd put visitors and evaluators together.

**Comment by samredai** [comment]
@rdblue how does this look?

- **literals.py**
  - IntegerLiteral
  - BooleanLiteral
  - DecimalLiteral
  - ...
- **base.py**
  - Literal
  - Expression
  - ExpressionVisitor
  - BoundExpressionVisitor
  - Evaluator
  - Bound
  - Unbound
  - Term
  - Predicate
  - Operation
  - And
  - Or
  - Not
  - TrueExp
  - FalseExp
  - BoundPredicate
  - UnboundPredicate
  - Boundreference
  - UnboundReference
- **visitors.py**
  - EvalVisitor
  - BindVisitor
  - ReferenceVisitor
  - ManifestEvalVisitor
  - MetricsEvalVisitor
  - InclusiveManifestEvaluator
  - StrictMetricsEvaluator

**Comment by Fokko** [comment]
@dramaticlly I think this one is not relevant anymore. We changed the package name to `pyiceberg`, and removed the src class. Next, the base.py is merged into each module's `__init__.py`. I'm closing this one if that's okay with you. Feel free to re-open if you want to reiterate the structure. Thanks!
